middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -1413,6 +1410,75 @@ let close_functions acc external_env ~current_region function_declarations =
       Ident.Map.empty func_decl_list
   in
+  let function_code_ids =
+    List.fold_left
+      (fun map decl ->
+        let function_slot = Function_decl.function_slot decl in
+        let code_id =
+          Code_id.create
+            ~name:(Function_slot.to_string function_slot)
+            compilation_unit
+        in
+        Function_slot.Map.add function_slot code_id map)
+      Function_slot.Map.empty func_decl_list
+  in
+  let approx_map =
+    List.fold_left
+      (fun approx_map decl ->
+        (* The only fields of metadata which are used for this pass are
+           params_arity, is_tupled, num_trailing_local_params,
+           contains_no_escaping_local_allocs, and result_arity. We try to
+           populate the different fields as much as possible, but put dummy
+           values when they are not yet computed or simply too expensive to
+           compute for the other fields. *)
+        let function_slot = Function_decl.function_slot decl in
+        let code_id = Function_slot.Map.find function_slot function_code_ids in
+        let params = Function_decl.params decl in
+        let params_arity =
+          List.map
+            (fun (_, kind) -> Flambda_kind.With_subkind.from_lambda kind)
+            params
+          |> Flambda_arity.With_subkinds.create
+        in
+        let return = Function_decl.return decl in
+        let result_arity =
+          Flambda_arity.With_subkinds.create [K.With_subkind.from_lambda return]
+        in
+        let check =
+          Check_attribute.from_lambda (Function_decl.check_attribute decl)
+        in
+        let cost_metrics = Cost_metrics.zero in
+        let dbg = Debuginfo.from_location (Function_decl.loc decl) in
+        let is_tupled =
+          match Function_decl.kind decl with
+          | Curried _ -> false
+          | Tupled -> true
+        in
+        let metadata =
+          Code_metadata.create code_id ~params_arity
+            ~num_trailing_local_params:
+              (Function_decl.num_trailing_local_params decl)
+            ~result_arity ~result_types:Unknown
+            ~contains_no_escaping_local_allocs:
+              (Function_decl.contains_no_escaping_local_allocs decl)
+            ~stub:(Function_decl.stub decl) ~inline:Never_inline ~check
+            ~is_a_functor:(Function_decl.is_a_functor decl)
+            ~recursive:(Function_decl.recursive decl)
+            ~newer_version_of:None ~cost_metrics
+            ~inlining_arguments:(Inlining_arguments.create ~round:0)
+            ~dbg ~is_tupled ~is_my_closure_used:true
+            ~inlining_decision:Recursive
+            ~absolute_history:(Inlining_history.Absolute.empty compilation_unit)
+            ~relative_history:Inlining_history.Relative.empty
+        in
+        let code = Code_or_metadata.create_metadata_only metadata in
+        let approx =
+          Value_approximation.Closure_approximation
+            { code_id; function_slot; code; symbol = None }
```

Comment:
I mean here. But I assume it would be useful only in a fonction that refer to itself in some way other than an application, so fairly limited.